core/src/main/scala/fs2/job/JobManager.scala

```diff
@@ -0,0 +1,214 @@
+        val killer = Context[F](Status.Canceled, None)
+        Concurrent[F].delay(meta.replace(id, old, killer)) flatMap { success =>
+          if (success)
+            Concurrent[F].unit
+          else
+            cancel(id)
+        }
+
+      case _ => Concurrent[F].unit
+    }
+  }
+
+  /**
+   * Returns the status of a given job id, if known.
+   */
+  def status(id: I): F[Option[Status]] =
+    Concurrent[F].delay(Option(meta.get(id)).map(_.status))
+
+  private def shutdown: F[Unit] =
+    Concurrent[F].delay(meta.clear()) >> notificationsQ.enqueue1(None)
+
+  private[this] def managementMachinery[A](
+      id: I,
+      in: Stream[F, A],
+      ignoreAbsence: Boolean): Stream[F, A] = {
+    Stream force {
+      SignallingRef[F, Boolean](false) map { s =>
+        lazy val frontF: F[Unit] = {
+          Concurrent[F].delay(Option(meta.get(id))) flatMap {
+            case Some(old @ Context(Status.Pending, _)) =>
+              val casF = Concurrent[F] delay {
+                meta.replace(id, old, Context(Status.Running, Some(s.set(true))))
+              }
+
+              casF flatMap { result =>
+                if (result)
+                  Concurrent[F].unit
+                else
+                  frontF
+              }
+
+            case Some(Context(Status.Canceled, _)) =>
+              Concurrent[F].delay(meta.remove(id)).void
+
+            case Some(Context(Status.Running, _)) =>
+              Concurrent[F].raiseError(new RuntimeException("job already running!"))    // TODO
+
+            case None =>
+              if (ignoreAbsence) {
+                val casF = Concurrent[F] delay {
+                  val attempt = Context(Status.Running, Some(s.set(true)))
+                  attempt eq meta.putIfAbsent(id, attempt)
+                }
+
+                casF flatMap { result =>
+                  if (result)
+                    Concurrent[F].unit
+                  else
+                    frontF
+                }
+              } else {
+                Concurrent[F].unit
+              }
+          }
+        }
+
+        Stream.bracket(frontF)(_ => Concurrent[F].delay(meta.remove(id)).void) >>
+          in.interruptWhen(s)
+      }
+    }
+  }
+}
+
+object JobManager {
+
+  @SuppressWarnings(Array("org.wartremover.warts.DefaultArguments"))
+  def apply[F[_]: Concurrent: Timer, I, N](
+      jobLimit: Int = 100,
+      notificationsLimit: Int = 10)   // all numbers are arbitrary, really
+      : Stream[F, JobManager[F, I, N]] = {
+
+    for {
+      notificationsQ <- Stream.eval(Queue.bounded[F, Option[N]](notificationsLimit))
+      dispatchQ <- Stream.eval(Queue.bounded[F, Stream[F, Nothing]](jobLimit))
+
+      initF = Concurrent[F] delay {
+        new JobManager[F, I, N](
+          notificationsQ,
+          dispatchQ)
+      }
+
+      jm <- Stream.bracket(initF)(_.shutdown)
+      back <- Stream.emit(jm).concurrently(dispatchQ.dequeue.parJoin(jobLimit))
```

Comment:
Won't this interrupt all running jobs if any of them fail?

Reply:
Ah yes it does! I need to handle errors and turn them into notifications.

Reply:
I'm not very familiar with concurrent streams, but from reading the documentation of `concurrently`, doesn't this mean that once we finish emitting `jm`, everything from `dispatchQ.dequeue.parJoin(jobLimit)` will halt? When will `Stream.emit(jm)` halt?